README.md
# Simple TCP socket proxy

This proxy was originally designed to allow a user to ssh into a Ubuntu multipass VM from the host machine or the Internet.

I am amazed about the simplicity and performance of the proxy, thus polishing it to become a generic socket proxy that you can use to do network address translation (NAT) from a host port to a guest VM target IP and port.

## proxy-config.json

The IP discovery command, source and target ports are configurable using a JSON file like this. You may adjust the configuration file to fit your use case.

```json
{
  "discovery": {
    "command": "multipass exec main ifconfig eth0",
    "tag": "inet",
    "index": 1
  },
  "source_port": [22],
  "target_port": [22],
  "authorized": ["192.168.1.*"]
}
```

## SSH security

If you use this to expose your guest VM to the Internet, make sure you use certificate authentication and disable password authentication. You should also use SSH certificate of 4,096 bits for security reason.

If you are using Multipass, you can replace the original SSH certificate with a 4,096-bit certificate.

## Running this utility as a service

We recommend that you use Node's PM2 process manager to deploy the utility as a service.

It is as simple as:

```
pm2 start simple-proxy.js
```
This assumes you start the utility from the project directory. If not, please add path to the filename.

## Library dependencies

Please install the moment time utility before you start the utility.

```
cd simple-proxy
npm install
```

## Port-Forward utility

A convenient port-forward utility is derived from the simple-proxy app.
This port-forward utility is particularly useful when using VMs in your laptop.
It allows you to reach the VM's applications (docker/kubernetes, etc.) from the host OS.

The port-forward utility is a command line tool. You can run it like this:

```
node port-forward.js source_ip:port target_ip:port
```
To listen to all IP addresses of the host, you may use 0.0.0.0 as the source_ip


## Idle disconnect timer

The system has an idle disconnect timeout value of 30 minutes, you may adjust it according to your use cases.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---
## Version 1.1.2, 4/2/2022

### Added

N/A

### Removed

N/A

### Changed

Minor refactoring

---
## Version 1.1.1, 11/25/2021

### Added

The "port forward" utility is derived from the Simple Proxy app

### Removed

N/A

### Changed

Set idle disconnect timer to 30 minutes

---
## Version 1.1.0, 8/5/2021

### Added

Support relay of multiple ports

### Removed

N/A

### Changed

N/A

---
## Version 1.0.0, 9/12/2020

### Added

First release

### Removed

N/A

### Changed

N/A


port-forward.js
const process = require('process');
const DateFormat = require('moment');
const Net = require('net');
const Crypto = require('crypto');
const NumberFormat = new Intl.NumberFormat();

const TIMESTAMP = 'YYYY-MM-DD HH:mm:ss.SSS';
const ZEROES = '000000';
const instanceId = (ZEROES + Crypto.randomBytes(4).readUIntBE(0, 4) % 10000).slice(-4);
const IDLE_TIMEOUT = 1800 * 1000;   // 30 minutes
var stopping = false;

const consoleLog = (message) => {
    console.log(DateFormat().format(TIMESTAMP)+' ['+instanceId+'] '+message);
};

const forwardPort = (sourceIp, sourcePort, targetIp, targetPort) => {
    const connections = new Map();
    // Setup TCP socket server
    const server = new Net.Server();
    server.listen(sourcePort, sourceIp, () => {
        consoleLog('Forwarding '+sourceIp+":"+sourcePort+' to '+targetIp+':'+targetPort);
    });
    // Graceful shutdown
    const gracefulShutdown = () => {
        for (const k of connections.keys()) {
            consoleLog('Stopping '+k);
            connections.get(k).end();
        }
        server.close(() => {
            consoleLog('Proxy '+sourceIp+":"+sourcePort+' to '+targetIp+':'+targetPort+' stopped');
        });
    };
    process.on('SIGTERM', () => {
        if (!stopping) {
            stopping = true;
            consoleLog('Kill signal detected');
        }
        gracefulShutdown();
    });
    process.on('SIGINT', () => {
        if (!stopping) {
            stopping = true;
            consoleLog('Control-C detected');
        }
        gracefulShutdown();
    });
    // Start TCP socket server
    server.on('connection', (socket) => {
        // check remote address
        const remoteIp = socket.remoteAddress;
        var normal = true;
        const sessionId = (ZEROES + Crypto.randomBytes(4).readUIntBE(0, 4) % 1000000).slice(-6);
        const client = Net.connect(targetPort, targetIp, () => {
            connections.set(sessionId, client);
            consoleLog( 'Session ' + sessionId + ' ' + remoteIp + ' connected to ' + targetIp + ':'+targetPort);
            server.getConnections((err, count) => {
                if (err) {
                    consoleLog(err.message);
                } else {
                    consoleLog("Total connections = " + count);
                }
            });
        });
        client.setTimeout(IDLE_TIMEOUT, () => {
            consoleLog('Session ' + sessionId + ' timeout');
            socket.end();
        });
        socket.on('data', (data) => {
            if (normal) client.write(data);
        });
        socket.once('end', () => {
            connections.delete(sessionId);
            consoleLog('Session '+sessionId+' closed by '+remoteIp);
            client.end();
            server.getConnections((err, count) => {
                if (err) {
                    consoleLog(err.message);
                } else {
                    consoleLog("Remaining connections = " + count);
                }
            });
        });
        client.on('data', (data) => {
            if (normal) socket.write(data);
        });
        client.once('end', () => {
            connections.delete(sessionId);
            consoleLog( 'Session '+sessionId+ ' ' + remoteIp + ' disconnected from ' + targetIp + ':' + targetPort +
                        ' rx ' + NumberFormat.format(socket.bytesRead) +
                        ' tx ' + NumberFormat.format(socket.bytesWritten));
            socket.end();
            server.getConnections((err, count) => {
                if (err) {
                    consoleLog(err.message);
                } else {
                    consoleLog("Remaining connections = " + count);
                }
            });
        });
        // Socket exceptions - most likely to be read timeout or connection reset
        socket.on('error', (err) => {
            if ('ECONNRESET' == err.code) {
                // normal case when user is using Windows
                consoleLog('Session '+sessionId+' closed by '+remoteIp);
            } else {
                consoleLog('Session '+sessionId+' exception ('+remoteIp+') - '+err.code);
            }
            normal = false;
            client.end();
            socket.end();
        });
        client.on('error', (err) => {
            consoleLog('Exception for port-'+targetPort+' - '+err);
            normal = false;
            socket.end();
            client.end();
        });
    });
}

function main() {
    const args = process.argv.slice(2);
    if (args.length == 2) {
        var source = get_address(args[0]);
        var target = get_address(args[1]);
        if (source != null && target != null) {
            forwardPort(source[0], source[1], target[0], target[1]);
            return;
        }
    }
    console.log("Usage: node port-forward.js source_ip:port target_ip:port");
}

function get_address(ip) {
    const colon = ip.lastIndexOf(':');
    if (colon > 2) {
        var result = [];
        result.push(ip.substring(0, colon));
        const port = ip.substring(colon+1);
        if (!isNaN(port)) {
            result.push(parseInt(port));
            return result;
        }
    }
    return null;
}

main()


simple-proxy.js
const Shell = require('child_process');
const DateFormat = require('moment');
const Net = require('net');
const Crypto = require('crypto');
const Path = require('path');
const Fs = require('fs');
const NumberFormat = new Intl.NumberFormat();

const APP_NAME = 'Simple Proxy v1.1.2';
const TIMESTAMP = 'YYYY-MM-DD HH:mm:ss.SSS';
const ZEROES = '000000';
const instanceId = (ZEROES + Crypto.randomBytes(4).readUIntBE(0, 4) % 10000).slice(-4);
const IDLE_TIMEOUT = 1800 * 1000;   // 30 minutes
var stopping = false;

const consoleLog = (message) => {
    console.log(DateFormat().format(TIMESTAMP)+' ['+instanceId+'] '+message);
};

async function portReady(host, port) {
    return new Promise((resolve, reject) => {
        const client = Net.createConnection({ port: port, host: host }, () => {
            client.end();
            resolve(true);
        });
        client.on('error', function(err) {
            client.end();
            resolve(false);
        });
    });
};

async function getVmIpAddress(command, key, index) {
    return new Promise((resolve, reject) => {
        Shell.exec(command, (error, stdout, stderr) => {
            if (error) {
                consoleLog(error.message.split('\n')[0]);
            } 
            if (stdout) {
                var ip = stdout.split('\n').filter(v => v.toString().trim().startsWith(key+' '))[0].trim().split(' ').filter(v => v.length > 0)[index];
                resolve([true, ip]);
            } else {
                resolve([false, stderr.split('\n')[0]]);
            }
        });
    });
}

const isAuthorized = (remoteIp, authorized) => {
    if (authorized.includes(remoteIp)) {
        return true;
    }
    for (i in authorized) {
        const s = authorized[i];
        if (s.endsWith('.*')) {
            const prefix = s.substring(0, s.lastIndexOf('.'));
            if (remoteIp.startsWith(prefix+'.')) {
                return true;
            }
        }
    }
    return false;
}

const forwardPort = (sourcePort, targetIp, targetPort, authorized, restart) => {
    const connections = new Map();
    // Setup TCP socket server
    const server = new Net.Server();
    server.listen(sourcePort, '0.0.0.0', () => {
        consoleLog('Forwarding port-'+sourcePort+' to '+targetIp+':'+targetPort);
    });
    // Graceful shutdown
    const gracefulShutdown = () => {
        for (const k of connections.keys()) {
            consoleLog('Stopping '+k);
            connections.get(k).end();
        }
        server.close(() => {
            consoleLog('Proxy service port-'+targetPort+' stopped');
        });
    };
    process.on('SIGTERM', () => {
        if (!stopping) {
            stopping = true;
            consoleLog('Kill signal detected');
        }
        gracefulShutdown();
    });
    process.on('SIGINT', () => {
        if (!stopping) {
            stopping = true;
            consoleLog('Control-C detected');
        }
        gracefulShutdown();
    });
    // Start TCP socket server
    server.on('connection', (socket) => {
        // check remote address
        const remoteIp = socket.remoteAddress;
        if (!isAuthorized(remoteIp, authorized)) {
            consoleLog("Unknown caller " + remoteIp + " connection to " + targetPort + " rejected");
            socket.destroy();
            return;
        }
        var normal = true;
        const sessionId = (ZEROES + Crypto.randomBytes(4).readUIntBE(0, 4) % 1000000).slice(-6);
        const client = Net.connect(targetPort, targetIp, () => {
            connections.set(sessionId, client);
            consoleLog( 'Session ' + sessionId + ' ' + remoteIp + ' connected to ' + targetIp + ':'+targetPort);
            server.getConnections((err, count) => {
                if (err) {
                    consoleLog(err.message);
                } else {
                    consoleLog("Total connections = " + count);
                }
            });
        });
        client.setTimeout(IDLE_TIMEOUT, () => {
            consoleLog('Session ' + sessionId + ' timeout');
            socket.end();
        });
        socket.on('data', (data) => {
            if (normal) client.write(data);
        });
        socket.once('end', () => {
            connections.delete(sessionId);
            consoleLog('Session '+sessionId+' closed by '+remoteIp);
            client.end();
            server.getConnections((err, count) => {
                if (err) {
                    consoleLog(err.message);
                } else {
                    consoleLog("Remaining connections = " + count);
                }
            });
        });
        client.on('data', (data) => {
            if (normal) socket.write(data);
        });
        client.once('end', () => {
            connections.delete(sessionId);
            consoleLog( 'Session '+sessionId+ ' ' + remoteIp + ' disconnected from ' + targetIp + ':' + targetPort +
                        ' rx ' + NumberFormat.format(socket.bytesRead) +
                        ' tx ' + NumberFormat.format(socket.bytesWritten));
            socket.end();
            server.getConnections((err, count) => {
                if (err) {
                    consoleLog(err.message);
                } else {
                    consoleLog("Remaining connections = " + count);
                }
            });
        });
        // Socket exceptions - most likely to be read timeout or connection reset
        socket.on('error', (err) => {
            if ('ECONNRESET' == err.code) {
                // normal case when user is using Windows
                consoleLog('Session '+sessionId+' closed by '+remoteIp);
            } else {
                consoleLog('Session '+sessionId+' exception ('+remoteIp+') - '+err.code);
            }
            normal = false;
            client.end();
            socket.end();
        });
        client.on('error', (err) => {
            consoleLog('Exception for port-'+targetPort+' - '+err);
            // Caution: the error message may depend on locale and language
            if (err.message.startsWith('connect ETIMEDOUT') && targetPort == restart) {
                // Let process manager to restart this app
                consoleLog('Stopping application because port-'+targetPort+' does not respond');
                process.exit(1);
            }
            normal = false;
            socket.end();
            client.end();
        });
    });
}

async function main() {
    var fileName = Path.resolve(__dirname, 'proxy-config.json');
    consoleLog(APP_NAME);
    consoleLog('Loading config from '+fileName);
    if (Fs.existsSync(fileName)) {
        const fd = Fs.openSync(fileName);
        const text = Fs.readFileSync(fd, 'utf-8');
        Fs.closeSync(fd);
        const json = JSON.parse(text);
        const command = json['discovery']['command'];
        const tag = json['discovery']['tag'];
        const index = json['discovery']['index'];
        const authorized = json['authorized'];
        const restart = json['restart'];
        consoleLog("Authorized users "+JSON.stringify(authorized));
        const source_ports = json['source_ports'];
        const target_ports = json['target_ports'];
        if (command && tag && index && source_ports && target_ports) {
            if (source_ports.length != target_ports.length) {
                consoleLog('Invalid proxy-config.json');
            } else {
                // Obtain dynamic IP address - this assumes we are using multipass and the VM is called "main"
                const [valid, targetIp] = await getVmIpAddress(command, tag, index);
                if (!valid) {
                    consoleLog('Unable to obtain target IP address - '+targetIp);
                } else {
                    for (i in source_ports) {
                        forwardPort(source_ports[i], targetIp, target_ports[i], authorized, restart);
                    }
                }
            }
        } else {
            consoleLog('Invalid proxy-config.json');
        }
    } else {
        consoleLog('Missing proxy-config.json');
    }
}

// start main application
main();
